### PR TITLE
Ticket/15253 new

### PR DESCRIPTION
**Pull request by rubencm**
Checklist:

- [ ] Correct branch: master for new features; 3.2.x, 3.1.x for fixes
- [ ] Tests pass
- [ ] Code follows coding guidelines: [master / 3.2.x](https://area51.phpbb.com/docs/master/coding-guidelines.html), [3.1.x](https://area51.phpbb.com/docs/31x/coding-guidelines.html)
- [ ] Commit follows commit message [format](https://wiki.phpbb.com/Git#Commit_Messages)

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-12345

### CODE REVIEW

**Comment by rubencm** [comment]
@Nicofuma done

**Comment by Nicofuma** [comment]
thanks

**Comment by Nicofuma** [comment]
Okay, found the good branch I think

File: phpBB/phpbb/storage/provider/provider_interface.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+namespace phpbb\storage\provider;
+
+interface provider_interface
+{
+	/**
+	 * Gets adapter class.
+	 *
+	 * @return \phpbb\storage\adapter\adapter_interface
+	 */
+	public function get_class();
```
Comment (Nicofuma) [inline phpBB/phpbb/storage/provider/provider_interface.php]: you should rename it to get_adapter_class() I think

File: phpBB/phpbb/storage/storage.php [hunk 2]
```diff
@@ -18,43 +18,110 @@
...
+	/**
+	 * Constructor
+	 *
+	 * @param \phpbb\storage\adapter_factory	$factory
+	 * @param string							$storage_name
+	 */
+	public function __construct(adapter_factory $factory, $storage_name)
 	{
-		$this->adapter = $adapter;
+		$this->adapter = $factory->get($storage_name);
```
Comment (Nicofuma) [inline phpBB/phpbb/storage/storage.php]: you should do lazily: store the factory in a property, adds a protected get_adapter() method and everywhere in the file uses get_adapter().

get_adapter:
```
protected function get_adapter()
{
    if($this->adapter === null) {
        $this->adapter = $this->factory->get($this->storage_name);
    }

    return $this->adapter;
}
```

File: phpBB/phpbb/storage/storage.php [hunk 3]
```diff
@@ -18,43 +18,110 @@
...
 	}
 
+	/**
+	 * Creates a directory recursively.
+	 *
+	 * @param string	$path	The directory path
+	 *
+	 * @throws \phpbb\storage\exception\exception	On any directory creation failure
+	 */
 	public function create_dir($path)
```
Comment (Nicofuma) [inline phpBB/phpbb/storage/storage.php]: do we keep this function?

**Comment by Nicofuma** [comment]
pretty good :)

File: phpBB/phpbb/storage/storage.php (hunk 3 above)
Comment (rubencm) [reply phpBB/phpbb/storage/storage.php]: Depending on the storage system, maybe is necessary

Comment (Nicofuma) [reply phpBB/phpbb/storage/storage.php]: but not from the outside, right? I think it's the role of the adapter to deal with directories. The class using a storage shouldn't care about how it is stored.

Comment (rubencm) [reply phpBB/phpbb/storage/storage.php]: yes, from inside, i can make it protected

Comment (Nicofuma) [reply phpBB/phpbb/storage/storage.php]: but inside the storage class? for me only the adapter may need to create a directory

Comment (rubencm) [reply phpBB/phpbb/storage/storage.php]: inside the adapter

Comment (rubencm) [reply phpBB/phpbb/storage/storage.php]: ok, i will change it

**Comment by rubencm** [comment]
Thank you

File: phpBB/phpbb/storage/storage.php (hunk 2 above)
Comment (rubencm) [reply phpBB/phpbb/storage/storage.php]: This is a very good idea, and solves a problem that i already knew i was going to have

**Comment by Nicofuma** [comment]
I think you should close this PR as you have it open against phpbb/phpbb

**Pull request closed without merging**